src/util/requests/RequestBuilder.cpp

```diff
@@ -0,0 +1,205 @@
+#include <boost/beast/http/field.hpp>
+#include <boost/beast/http/message.hpp>
+#include <boost/beast/http/status.hpp>
+#include <boost/beast/http/string_body.hpp>
+#include <boost/beast/http/verb.hpp>
+#include <boost/beast/ssl/ssl_stream.hpp>
+#include <boost/beast/version.hpp>
+#include <openssl/err.h>
+#include <openssl/tls1.h>
+
+#include <chrono>
+#include <string>
+#include <string_view>
+#include <utility>
+#include <vector>
+
+namespace util::requests {
+
+namespace beast = boost::beast;
+namespace http = beast::http;
+namespace asio = boost::asio;
+using tcp = asio::ip::tcp;
+
+RequestBuilder::RequestBuilder(std::string host, std::string port) : host_(std::move(host)), port_(std::move(port))
+{
+    request_.set(http::field::host, host_);
+    request_.target("/");
+}
+
+RequestBuilder&
+RequestBuilder::addHeader(HttpHeader const& header)
+{
+    request_.set(header.name, header.value);
+    return *this;
+}
+
+RequestBuilder&
+RequestBuilder::addHeaders(std::vector<HttpHeader> const& headers)
+{
+    for (auto const& header : headers)
+        addHeader(header);
+    return *this;
+}
+
+RequestBuilder&
+RequestBuilder::addData(std::string data)
+{
+    request_.body() = data;
+    request_.prepare_payload();
+    return *this;
+}
+
+RequestBuilder&
+RequestBuilder::setTimeout(std::chrono::milliseconds const timeout)
+{
+    timeout_ = timeout;
+    return *this;
+}
+
+RequestBuilder&
+RequestBuilder::setTarget(std::string_view target)
+{
+    request_.target(target);
+    return *this;
+}
+
+RequestBuilder&
+RequestBuilder::setSslEnabled(bool const enabled)
+{
+    sslEnabled_ = enabled;
+    return *this;
+}
+
+Expected<std::string, RequestError>
+RequestBuilder::get(asio::yield_context yield)
+{
+    return doRequest(yield, http::verb::get);
+}
+
+Expected<std::string, RequestError>
+RequestBuilder::post(asio::yield_context yield)
+{
+    return doRequest(yield, http::verb::post);
+}
+
+Expected<std::string, RequestError>
+RequestBuilder::doRequest(asio::yield_context yield, beast::http::verb method)
+{
+    if (sslEnabled_) {
+        auto streamData = impl::SslTcpStreamData::create(yield);
+        if (not streamData.has_value())
+            return Unexpected{std::move(streamData.error())};
+
+        if (!SSL_set_tlsext_host_name(streamData->stream.native_handle(), host_.c_str())) {
+            beast::error_code errorCode;
+            errorCode.assign(static_cast<int>(::ERR_get_error()), asio::error::get_ssl_category());
+            return Unexpected{RequestError{"SSL setup failed", errorCode}};
+        }
+        return doRequestImpl(std::move(streamData.value()), yield, method);
+    }
+
+    auto streamData = impl::TcpStreamData{yield};
+    return doRequestImpl(std::move(streamData), yield, method);
+}
+
+template <typename StreamDataType>
+Expected<std::string, RequestError>
+RequestBuilder::doRequestImpl(StreamDataType&& streamData, asio::yield_context yield, http::verb const method)
```

Comment:
Do we need && or just & here.

Reply:
Hmm, it will work fine with `&`. I thought moving stream is a more safe design for future changes, but maybe there is no difference. I'm ok with both options.Do you think I should change it?

Reply:
I personally will change it to &. Because the implementation does not need it being right value for now. But maybe future.